js/uid2-sdk-3.0.0/callbacks.test.ts

```diff
@@ -0,0 +1,132 @@
+  jest.runOnlyPendingTimers();
+
+  callback = jest.fn();
+  uid2 = new UID2();
+  xhrMock = new mocks.XhrMock(sdkWindow);
+  _cryptoMock = new mocks.CryptoMock(sdkWindow);
+  mocks.setCookieMock(sdkWindow.document);
+  sdkWindow.__uid2 = { callbacks: [] };
+  asyncCallback = jest.fn((event, payload) => {
+    console.log("Async Callback Event:", event);
+    console.log("Payload:", payload);
+  });
+});
+
+afterEach(() => {
+  mocks.resetFakeTime();
+});
+
+const makeIdentity = mocks.makeIdentityV2;
+
+describe("when a callback is provided", () => {
+  const refreshFrom = Date.now() + 100;
+  const identity = { ...makeIdentity(), refresh_from: refreshFrom };
+  const refreshedIdentity = { ...makeIdentity(), advertising_token: 'refreshed_token' };
+  describe("before init is called", () => {
+    test("it should be called at the end of init", () => {
+      uid2.callbacks.push(asyncCallback);
+      const calls = asyncCallback.mock.calls.length;
+      uid2.init({ callback: callback, identity: identity });
+      expect(asyncCallback.mock.calls.length).toBe(calls+1);
+    });
+    test("it should be called with a 'successful init' message", () => {
+      uid2.callbacks.push(asyncCallback);
+      uid2.init({ callback: callback, identity: identity });
+      expect(asyncCallback.mock.calls.slice(-1)[0][0]).toBe(UID2.EventType.InitCompleted);
+    });
+    test("it should be provided with the loaded identity", () => {
+      uid2.callbacks.push(asyncCallback);
+      uid2.init({ callback: callback, identity: identity });
+      console.log(asyncCallback.mock.calls.slice(-1)[0][1]);
+      expect(asyncCallback.mock.calls.slice(-1)[0][1]).toMatchObject({ identity });
+    });
+  });
+
+  describe("after init is called", () => {
+    test("it should be called with SdkLoaded and InitComplete immediately", () => {
+      uid2.init({ callback: callback, identity: identity });
+      expect(asyncCallback.mock.calls.length).toBe(0);
+      uid2.callbacks.push(asyncCallback);
+      expect(asyncCallback.mock.calls.length).toBe(2);
+      expect(asyncCallback.mock.calls[0][0]).toBe(UID2.EventType.SdkLoaded);
+      expect(asyncCallback.mock.calls[1][0]).toBe(UID2.EventType.InitCompleted);
+    });
+
+    test("it should be provided with the loaded identity", () => {
+      uid2.init({ callback: callback, identity: identity });
+      uid2.callbacks.push(asyncCallback);
+      
+      expect(asyncCallback.mock.calls.slice(-1)[0][1]).toMatchObject({ identity });
+    });
+
+    test("it should receive subsequent identity updates", async () => {
+      uid2.init({ callback: callback, identity: identity });
+      uid2.callbacks.push(asyncCallback);
+      
+      const callsBeforeRefresh = asyncCallback.mock.calls.length;
+      jest.setSystemTime(refreshFrom);
+      jest.runOnlyPendingTimers();
+      expect(xhrMock.send).toHaveBeenCalledTimes(1);
+      xhrMock.sendRefreshApiResponse(refreshedIdentity);
+      await mocks.flushPromises();
```

Comment:
what is the use of flushPromises here? why couldn't use a single Promise.resolve here?

Reply:
Something about the JavaScript event loop and the chain of operations happening - a single call to Promise.resolve doesn't let it finish. Two calls is enough (which is all flushPromises does) - I tried to find a more general solution but spent a bunch of time and didn't find a better option.

Reply:
got it, thanks